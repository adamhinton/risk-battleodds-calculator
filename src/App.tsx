// TODO: Refactor, particularly make the simulation more DRY and figure out how to best replicate userInputs in runSingleSimulation

function App() {
  return (
    <div className="App">
      <h1>Adam Hinton</h1>
    </div>
  );
}

function generatePlayerRolls (playerType: PlayerType, diceRolls: 1 | 2 | number): PlayerRolls {
  const generateSortedRolls = (diceRolls: 1 | 2 | 3): PlayerRolls =>{
    const solution = [] 
    for(let i=0; i<diceRolls; i++){
      solution.push(randomIntFromInterval(1, 6))
    }
    // TODO: Refactor this, I shouldn't need this type assertion
   return sortPlayerRolls(solution)
  }

  // if one unit, doesn't matter if attacker or defender, they get one dice roll
  if(diceRolls === 1){
    return generateSortedRolls(1)
  }

  // They're a defender and we already know they don't have one unit, so they get two dice rolls
  if(playerType === 'defender'){
    return generateSortedRolls(2)
  }

  // We know they're an attacker now. So if they have two units they get two rolls, otherwise they get three rolls
  return diceRolls === 2? generateSortedRolls(2) : generateSortedRolls(3)
}

export default App;

// TYPE DEFINITIONS
type PlayerCounts = {
  attackerCount: number;
  defenderCount: number;
}

type Results = {
  attackerOccupies: number;
  defenderHolds: number;
  // averageUnitsLeft: {
  //   attackers: number;
  //   defenders: number;
  // }
}

type UserInputs = {
  playerCounts: PlayerCounts;
  numSimulations: number;
}

type PlayerType = 'attacker' | 'defender'

type AttackerRolls = [number, number? , number?]
type DefenderRolls = [number, number?]
type PlayerRolls = AttackerRolls | DefenderRolls

// USEFUL VARIABLES
const userInputs: UserInputs = {
  playerCounts: {
   attackerCount: 20,
   defenderCount: 20
  },
  numSimulations: 1
} 

const results: Results = {
  attackerOccupies: 0,
  defenderHolds: 0
}

// UTILS
const randomIntFromInterval = (min: Readonly<number>, max: Readonly<number>) =>{ // min and max included 
  return Math.floor(Math.random() * (max - min + 1) + min)
}

const sortPlayerRolls = (rolls: PlayerRolls) =>{
  return rolls.sort((a , b ) =>{ return b!  - a! })
}

console.time('a')
for(let i=0; i<userInputs.numSimulations; i++){
  runSingleSimulation( {...userInputs.playerCounts } )
}

console.log('results:', results)
console.timeEnd('a')
console.log('userInputs:', userInputs)

function runSingleSimulation (playerCounts: PlayerCounts): void {

// One run through of simulation
while ( playerCounts.attackerCount > 0 && playerCounts.defenderCount > 0 ){

const attackerRolls = generatePlayerRolls('attacker', playerCounts.attackerCount)
const defenderRolls = generatePlayerRolls('defender', playerCounts.defenderCount)

console.log('attackerRolls, defenderRolls:', attackerRolls, defenderRolls)

// const attackerRolls = sortPlayerRolls(attackerRolls)
// const defenderRollsSorted = sortPlayerRolls(defenderRolls)

// if diceRolls is 1 or 2 then follow certain logic. If more than 2 (aka its type is number) then follow other logic.


// Now for the attack 
if(attackerRolls[0] > defenderRolls[0]){
  playerCounts.defenderCount --
}
else {
  playerCounts.attackerCount --
}

if(playerCounts.attackerCount > 1 && playerCounts.defenderCount > 1 ){
  if(attackerRolls[1]! > defenderRolls[1]!){
  playerCounts.defenderCount --
}
  else {
  playerCounts.attackerCount --
   }
}

// results
if(playerCounts.attackerCount === 0){
  results.defenderHolds ++
}
else if (playerCounts.defenderCount === 0){
  results.attackerOccupies ++
}
}
}