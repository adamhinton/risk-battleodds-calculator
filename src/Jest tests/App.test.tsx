import { render, screen, fireEvent } from "@testing-library/react";
import App from "../App";
import renderer from "react-test-renderer";

// APP.TSX UNIT TESTS
test("[1] Sanity check", () => {
	expect(1 + 1).toEqual(2);
});

test("[2] renders without errors", () => {
	render(<App />);

	expect(screen.getByText(/run simulations/i)).toBeInTheDocument();
});

test("[3] Matches snapshot from 1.29.2023", () => {
	const tree = renderer.create(<App />).toJSON();
	expect(tree).toMatchSnapshot();
});

test("[4] Dark mode toggle button changes styling", () => {
	render(<App />);

	const darkModeButton = screen.getByTestId("darkmode-btn");

	expect(screen.getByTestId("app")).toHaveStyle(
		"background-color: rgb(45, 60, 66)"
	);

	fireEvent.click(darkModeButton);

	expect(screen.getByTestId("app")).toHaveStyle("background-color: #3c9893");
});
