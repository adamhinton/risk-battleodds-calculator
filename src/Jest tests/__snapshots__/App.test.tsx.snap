// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[3] Matches snapshot from 1.29.2023 1`] = `
<div
  className="sc-fPXMVe kLbkjt App"
  data-testid="app"
>
  <header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic sc-jEACwC dvRpkg css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <h1
      className="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      data-testid="title"
      style={
        Object {
          "fontSize": "1.2rem",
          "fontWeight": 600,
        }
      }
    >
      <div>
        Risk Battleodds Calculator 
        <img
          className="sc-eldPxv fmVEHI"
          src="/risk-favicon.png"
        />
      </div>
    </h1>
    <div
      className="sc-cPiKLX iujfqc"
      data-testid="links"
    >
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways sc-eDPEul cZKbJC css-1bhi0za-MuiTypography-root-MuiLink-root"
        href="https://github.com/adamhinton/risk-battleodds-calculator"
        onBlur={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        Source Code
      </a>
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways sc-eDPEul cZKbJC css-1bhi0za-MuiTypography-root-MuiLink-root"
        href="https://www.linkedin.com/in/adam-hinton/"
        onBlur={[Function]}
        onFocus={[Function]}
        target="_blank"
      >
        LinkedIn
      </a>
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways sc-eDPEul cZKbJC css-1bhi0za-MuiTypography-root-MuiLink-root"
        href="https://www.dicebreaker.com/games/risk/how-to/how-to-play-risk-board-game#:~:text=In%20Risk%2C%20players%20are%20competing,attack%20and%20capture%20neighbouring%20territories"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        Risk Explained
      </a>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      style={
        Object {
          "fontSize": "0.8rem",
        }
      }
    >
      For the classic board game 
      <b>
        RISK
      </b>
    </p>
    <div
      className="sc-dLMFU iNZwtN"
    >
      <h2
        className="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        data-testid="author"
        style={
          Object {
            "fontSize": "1rem",
            "fontWeight": 600,
          }
        }
      >
        Author: Adam Hinton
      </h2>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="LightModeIcon"
        focusable="false"
        onClick={[Function]}
        style={
          Object {
            "cursor": "pointer",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zM2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1zm18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1zM11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1zm0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1zM5.99 4.58c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41L5.99 4.58zm12.37 12.37c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0 .39-.39.39-1.03 0-1.41l-1.06-1.06zm1.06-10.96c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06zM7.05 18.36c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0l1.06-1.06z"
        />
      </svg>
    </div>
  </header>
  <main
    className="sc-gFqAkR eaZPol"
  >
    <form
      className="sc-gsFSXq fdGnTB"
      onSubmit={[Function]}
    >
      <div
        className="sc-kAyceB cqDIOQ"
      >
        <h2>
          Battle Outcomes Calculator
        </h2>
      </div>
      <div
        className="sc-imWYAI iNhoGk"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
          htmlFor="attackers"
        >
          Attacking Troops
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary sc-jXbUNg dJkBOO css-q0jhri-MuiInputBase-root-MuiInput-root"
          onClick={[Function]}
        >
          <input
            className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="attackers"
            name="attackerCount"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="number"
            value={10}
          />
        </div>
      </div>
      <div
        className="sc-imWYAI iNhoGk"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
          htmlFor="defenders"
        >
          Defending Troops
          <svg
            aria-hidden={true}
            aria-label="Comma-separated list of defenders (e.g., 10,5,5)"
            aria-labelledby={null}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-mui-internal-clone-element={true}
            data-testid="QuestionMarkIcon"
            focusable="false"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            viewBox="0 0 24 24"
          >
            <path
              d="M11.07 12.85c.77-1.39 2.25-2.21 3.11-3.44.91-1.29.4-3.7-2.18-3.7-1.69 0-2.52 1.28-2.87 2.34L6.54 6.96C7.25 4.83 9.18 3 11.99 3c2.35 0 3.96 1.07 4.78 2.41.7 1.15 1.11 3.3.03 4.9-1.2 1.77-2.35 2.31-2.97 3.45-.25.46-.35.76-.35 2.24h-2.89c-.01-.78-.13-2.05.48-3.15zM14 20c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2z"
            />
          </svg>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary sc-jXbUNg dJkBOO css-q0jhri-MuiInputBase-root-MuiInput-root"
          onClick={[Function]}
        >
          <input
            className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="defenders"
            name="defenderCount"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value="10"
          />
        </div>
      </div>
      <div
        className="sc-imWYAI iNhoGk"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
          htmlFor="stop-at"
        >
          Stop At
          <svg
            aria-hidden={true}
            aria-label="Stop at this number of attackers left"
            aria-labelledby={null}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-mui-internal-clone-element={true}
            data-testid="QuestionMarkIcon"
            focusable="false"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            viewBox="0 0 24 24"
          >
            <path
              d="M11.07 12.85c.77-1.39 2.25-2.21 3.11-3.44.91-1.29.4-3.7-2.18-3.7-1.69 0-2.52 1.28-2.87 2.34L6.54 6.96C7.25 4.83 9.18 3 11.99 3c2.35 0 3.96 1.07 4.78 2.41.7 1.15 1.11 3.3.03 4.9-1.2 1.77-2.35 2.31-2.97 3.45-.25.46-.35.76-.35 2.24h-2.89c-.01-.78-.13-2.05.48-3.15zM14 20c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2z"
            />
          </svg>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary sc-jXbUNg dJkBOO css-q0jhri-MuiInputBase-root-MuiInput-root"
          onClick={[Function]}
        >
          <input
            className="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id="stop-at"
            name="stopAt"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="number"
            value={3}
          />
        </div>
      </div>
      <div
        className="sc-dhKdcB cFVwEv"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
          htmlFor="simulations"
        >
          Simulations
          <svg
            aria-hidden={true}
            aria-label="Number of simulations to run (max 10k)"
            aria-labelledby={null}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-mui-internal-clone-element={true}
            data-testid="QuestionMarkIcon"
            focusable="false"
            onBlur={[Function]}
            onFocus={[Function]}
            onMouseLeave={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            viewBox="0 0 24 24"
          >
            <path
              d="M11.07 12.85c.77-1.39 2.25-2.21 3.11-3.44.91-1.29.4-3.7-2.18-3.7-1.69 0-2.52 1.28-2.87 2.34L6.54 6.96C7.25 4.83 9.18 3 11.99 3c2.35 0 3.96 1.07 4.78 2.41.7 1.15 1.11 3.3.03 4.9-1.2 1.77-2.35 2.31-2.97 3.45-.25.46-.35.76-.35 2.24h-2.89c-.01-.78-.13-2.05.48-3.15zM14 20c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2z"
            />
          </svg>
        </label>
      </div>
      <label
        className="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd sc-jlZhew shYkk css-j204z7-MuiFormControlLabel-root"
      >
        <span
          className="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={null}
        >
          <input
            checked={true}
            className="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate={false}
            onChange={[Function]}
            required={false}
            type="checkbox"
          />
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          className="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Hide form after running simulations
        </span>
      </label>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium sc-dAlyuH bcABWL css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        Run Simulations
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        className="Toastify"
      />
    </form>
  </main>
</div>
`;
